### ISSUE TITLE
Finetuning CaffeNet 

### ISSUE THREAD

**Issue by revilokeb** [beniz/deepdetect#22]
@beniz: is there a straighforward way to finetune a CaffeNet using current deepdetect? I could not find in the API (http://www.deepdetect.com/api/#launch-a-training-job) anything related to "-weights" in the typical Caffe command line for finetuning 'caffe train -solver ... -weights ...'

**Comment by beniz** [comment]
@revilokeb there's no dedicated way yet through the API because my understanding is that it should not be so hard to do. This has been on my todolist though, so this issue is to be considered.

My understanding on how to do is as follows:

- copy and change the .protoxt files (including deploy) from the original model in order to prepare the new model
- copy the weights from the original model repository into the new one
- train via API (using new repository as target) and with modified learning rate (etc...) as needed

Now, still my understanding, when using the `mlp` template (which I assume to not be a common case when learning from images, where finetuning has had the most proven track), simply copying the weights and using the `template` parameter and options via API should recreate a model with the novel number of classes.

**Comment by beniz** [comment]
Actually, I forgot to document that it is possible to pass the `weights` parameter to the `model` object at service creation.

**Comment by beniz** [comment]
Fixed the documentation so that the `weights` parameter appears among the options for the Caffe `parameters_mllib`, see http://www.deepdetect.com/api/?shell#create-a-service

@revilokeb I've successfully fine-tuned a Googlenet, so if this use-case is still useful for you, let me know if you have issues. Note that the prototxt needs to be edited by hand unfortunately at this points since the name of last fully connected layer of the net needs to be changed in order for Caffe to initialize it randomly (and re-learn it), see http://caffe.berkeleyvision.org/gathered/examples/finetune_flickr_style.html

**Comment by revilokeb** [comment]
@beniz Thanks for the update, I will have a look and report back if I am running into any issues

**Comment by beniz** [comment]
Added `finetuning` boolean API keyword for Caffe a service creation, along the `weights` parameter. Allows the automatic preparation of a model template for finetuning.

I am closing this issue for now as finetuning is working very well in my tests on a variety of image classification tasks.

**Comment by neo01124** [comment]
Don't mean to open the issue but have a few questions about finetuning through deepdetect and in general.

- Is it possible to finetune only a certain number of higher layers and leave the first x layers untouched ? 
Somehow switch off the backpropogation for these x layers.
- Supposing I have a target dataset of 100k images where classes are similar to the imagenet dataset, does finetuning make more sense or training a linear classifier (SVM) using the output from the second last layer. I will be trying both out but any insight will be interesting and maybe helpful for other readers as well.

Also is there a deepdetect mailing list for such discussions ?
Thanks

**Comment by beniz** [comment]
>     Is it possible to finetune only a certain number of higher layers and leave the first x layers untouched ? Somehow switch off the backpropogation for these x layers (to prevent overfitting)

It is possible but cannot be controlled from the API. See http://caffe.berkeleyvision.org/gathered/examples/finetune_flickr_style.html and set the corresponding `lr_mult` values accordingly. FYI this is not in the API mostly because in my own experience finetuning the whole weights always yields better results, including when the risk of overfit is very high.

>     Supposing I have a target dataset of 100k images where classes are similar to the imagenet dataset, does finetuning make more sense or training a linear classifier (SVM) using the output from the second last layer. I will be trying both out but any insight will be interesting and maybe helpful for other readers as well.

See the very useful information on this page: http://cs231n.github.io/transfer-learning/

> Also is there a deepdetect mailing list for such discussions ?

[gitter](https://gitter.im/beniz/deepdetect) is the way we go, there's no in between for now. One reason is that China appears to be cut off from google groups.

**Comment by neo01124** [comment]
thanks again @beniz
